### PR TITLE
Added somebody groups management

### PR DESCRIPTION
**Pull request by Mnosek**
Still work in progress.

### CODE REVIEW

File: src/People/Api/PartialHandlers.cs [hunk 1]
```diff
@@ -245,6 +245,43 @@ internal class PartialHandlers : IHandlers {
 
                 return page;
             });
+
+            Handle.GET("/people/partials/groups", () => {
+                var p = new GroupsPage()
+                {
+                    Html = "/People/viewmodels/GroupsPage.html"
```
Comment (joozek78) [inline src/People/Api/PartialHandlers.cs]: why not move it to json?

File: src/People/Api/PartialHandlers.cs [hunk 2]
```diff
@@ -245,6 +245,43 @@ internal class PartialHandlers : IHandlers {
...
+                };
+
+                p.RefreshGroups();
```
Comment (joozek78) [inline src/People/Api/PartialHandlers.cs]: too many empty lines

File: src/People/Api/PartialHandlers.cs [hunk 3]
```diff
@@ -245,6 +245,43 @@ internal class PartialHandlers : IHandlers {
...
+            });
+
+            Handle.GET<string>("/people/partials/groups/{?}", (string id) => {
+                return Db.Scope<GroupPage>(() => {
+                    var page = new GroupPage()
+                    {
+                        Html = "/People/viewmodels/GroupPage.html"
+                    };
+
+                    page.RefreshGroup(id);
```
Comment (joozek78) [inline src/People/Api/PartialHandlers.cs]: if you make RefreshGroup return GroupPage itself you'll be down to one line:
`return Db.Scope<GroupPage>(() => new GroupPage().RefreshGroup(id))`

File: src/People/Helpers/GroupsProvider.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+using Starcounter;
+using Simplified.Ring2;
+
+namespace People
+{
+    public class GroupsProvider
+    {
+        public QueryResultRows<SomebodyGroup> SelectGroups()
+        {
+            return Db.SQL<SomebodyGroup>("SELECT sg FROM Simplified.Ring2.SomebodyGroup sg ORDER BY sg.Name");
```
Comment (joozek78) [inline src/People/Helpers/GroupsProvider.cs]: don't hardcode everything (or anything)

File: src/People/Helpers/GroupsProvider.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        public void Delete(SomebodyGroup group)
+        {
+            var somebodyRelations =
+                Db.SQL<SomebodyGroupMemberSomebody>(
+                    $"SELECT sgms FROM {typeof(SomebodyGroupMemberSomebody).FullName} sgms WHERE sgms.ToWhat = ?", group);
```
Comment (joozek78) [inline src/People/Helpers/GroupsProvider.cs]: ToWhat has a synonym to make it more clear what happens

File: src/People/Helpers/GroupsProvider.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+            foreach (var item in somebodyRelations)
+            {
+                item.Delete();
+            }
+
+            foreach (var item in groupRelations)
+            {
+                item.Delete();
+            }
```
Comment (joozek78) [inline src/People/Helpers/GroupsProvider.cs]: you can remove one foreach by concatting the items

File: src/People/Helpers/GroupsProvider.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+            group.Delete();
+        }
+
+        public bool IsGroupContainingGroup(SomebodyGroup haystack, SomebodyGroup needle)
+        {
+            return haystack.GetSubgroups().Any(item => item.Key == needle.Key || IsGroupContainingGroup(item, needle));          
+        }
+
+        public static string GetGroupAddUri()
```
Comment (joozek78) [inline src/People/Helpers/GroupsProvider.cs]: why not make it a property? `public static string GroupAddUri => "/people/groups/add";`

File: src/People/Helpers/GroupsProvider.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            return "/people/groups/add";
+        }
+
+        public static string GetGroupUri(SomebodyGroup group)
```
Comment (joozek78) [inline src/People/Helpers/GroupsProvider.cs]:
```c#
public static string GetGroupUri(SomebodyGroup group) =>$"/people/groups/{group.Key}";
```
Doesn't your R# hint at that?

File: src/People/ViewModels/GroupMemberAddingPage.json.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+using Starcounter;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace People {
+    partial class GroupMemberAddingPage : Json {
+        public Action MemberAdded { get; set; }
+        public SomebodyGroup SomebodyGroupToAddTo;
+        protected GroupsProvider GroupsProvider = new GroupsProvider();
```
Comment (joozek78) [inline src/People/ViewModels/GroupMemberAddingPage.json.cs]: protected?

File: src/People/ViewModels/GroupMemberAddingPage.json.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            var searchTerm = $"%{search}%".ToLower();
+            Results.Data = SearchForSomebodyGroups(searchTerm).Concat(SearchForSomebodies(searchTerm));
+
+            foreach (var item in Results)
+            {
+                item.SelectResult = AssignMember;
+            }
+        }
+
+        private void AssignMember(GroupResultItem item)
```
Comment (joozek78) [inline src/People/ViewModels/GroupMemberAddingPage.json.cs]: You use 'add' in one place and 'assign' in other. Be consistent. Reading the name, I don't know what this method does

File: src/People/ViewModels/GroupMemberAddingPage.json.cs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            if (item.Somebody.Data != null)
+            {
+                new SomebodyGroupMemberSomebody
+                {
+                    WhatIs = item.Somebody.Data,
```
Comment (joozek78) [inline src/People/ViewModels/GroupMemberAddingPage.json.cs]: We have those synonyms for a reason

File: src/People/ViewModels/GroupMemberAddingPage.json.cs [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+                    ToWhat = SomebodyGroupToAddTo
+                };
+            }
+
+            if (item.SomebodyGroup.Data != null)
+            {
+                if (GroupsProvider.IsGroupContainingGroup(item.SomebodyGroup.Data, SomebodyGroupToAddTo))
+                {
+                    return;
```
Comment (joozek78) [inline src/People/ViewModels/GroupMemberAddingPage.json.cs]: I assume there will be some fair warning here? In the future?

File: src/People/ViewModels/GroupMemberAddingPage.json.cs [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+                .Select(item => new GroupResultItem { Key = item.Key, SomebodyGroup = { Data = item } });            
+        }
+    }
+
+    [GroupMemberAddingPage_json.Results]
+    partial class GroupResultItem : Json
+    {
+        public Action<GroupResultItem> SelectResult { get; set; }
+            
+        public void Handle (Input.Select action)
```
Comment (joozek78) [inline src/People/ViewModels/GroupMemberAddingPage.json.cs]: why would you add a space here? `Handle (`

File: src/People/ViewModels/GroupMembersListPage.json.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
+using System.Linq;
+using Simplified.Ring2;
+using Starcounter;
+
+namespace People {
+    partial class GroupMembersListPage : Json {
+        public SomebodyGroup SomebodyGroupListFor { get; set; }
```
Comment (joozek78) [inline src/People/ViewModels/GroupMembersListPage.json.cs]: What does it mean? Why not IBound?

File: src/People/ViewModels/GroupMembersListPage.json.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        public void RefreshMembers()
+        {            
+            Group.Data = SomebodyGroupListFor;
+            Group.ICanDestroy();
+        }
+
+        [GroupMembersListPage_json.Group]
+        partial class GroupMembersListItemPage : Json, IBound<SomebodyGroup>
+        {
+            public void ICanDestroy()
```
Comment (joozek78) [inline src/People/ViewModels/GroupMembersListPage.json.cs]: I just want to make sure this name will change :)

Comment (joozek78) [inline src/People/ViewModels/GroupMembersListPage.json.cs]: BTW It's a great artist: https://open.spotify.com/album/6VEduKINi4zglNImeRbEC0

File: src/People/ViewModels/GroupMembersListPage.json.cs [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+                foreach (var subgroup in Subgroups)
+                {
+                    subgroup.CanDelete = true;
+                }
+            }
+
+            protected override void OnData()
+            {
+                base.OnData();
+                foreach (var subgroup in Data.GetSubgroups().ToList())
```
Comment (joozek78) [inline src/People/ViewModels/GroupMembersListPage.json.cs]: ToList() is redundant here

Comment (joozek78) [inline src/People/ViewModels/GroupMembersListPage.json.cs]: And BTW it's strange... [this](https://github.com/StarcounterSamples/People/pull/48/files#diff-e289c130d23257c934e79cd05e291462R29) works, so wouldn't it here?

File: src/People/ViewModels/GroupPage.json.cs [hunk 17]
```diff
@@ -0,0 +1,45 @@
+using Starcounter;
+using Simplified.Ring2;
+
+namespace People {
+    partial class GroupPage : Json, IBound<SomebodyGroup> {
+        protected GroupsProvider GroupsProvider = new GroupsProvider();
+
+        public void RefreshGroup(string Id = null)
```
Comment (joozek78) [inline src/People/ViewModels/GroupPage.json.cs]: Why not make it two separate methods when they have completely different meaning? You either load an existing group or create one

File: src/People/ViewModels/GroupPage.json.cs [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+            if (string.IsNullOrEmpty(Id))
+            {
+                Data = new SomebodyGroup();
+            }
+            else
+            {
+                Data = DbHelper.FromID(DbHelper.Base64DecodeObjectID(Id)) as SomebodyGroup;
+            }
+
+            var groupMemberAddingPage = new GroupMemberAddingPage { SomebodyGroupToAddTo = Data };
```
Comment (joozek78) [inline src/People/ViewModels/GroupPage.json.cs]: Those things can probably live in OnData

File: src/People/ViewModels/GroupPage.json.cs [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+            var groupMembersListPage = new GroupMembersListPage { SomebodyGroupListFor = Data };
+
+            groupMemberAddingPage.MemberAdded = () =>
+            {
+                groupMembersListPage.RefreshMembers();
+            };
+            
+            groupMembersListPage.RefreshMembers();
+            GroupMemberAddingPage = groupMemberAddingPage;
```
Comment (joozek78) [inline src/People/ViewModels/GroupPage.json.cs]: Is there a reason you are deferring those assignments?

File: src/People/wwwroot/People/viewmodels/GroupMemberAddingPage.html [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+                </li>
+            </template>
+        </ul>
+    </template>
+    <script>
+        (function () {
+            var script = document._currentScript || document.currentScript;
+            var template = script.previousElementSibling;
+
+            template.selectResult = function (ev) {
```
Comment (joozek78) [inline src/People/wwwroot/People/viewmodels/GroupMemberAddingPage.html]: I'm just asking... is this required or is it artifact of copy-paste?

**Comment by Mnosek** [comment]
Thanks for feedback! Told you, that this is work in progress... :) I will definitely include your remarks.

File: src/People/Helpers/GroupsProvider.cs (hunk 5 above)
Comment (Mnosek) [inline src/People/Helpers/GroupsProvider.cs]: I've changed synonyms in my whole code

**Comment by Mnosek** [comment]
Fixed with your suggestions.

**Pull request merged**